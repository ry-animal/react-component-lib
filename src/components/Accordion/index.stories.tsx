import { ComponentStory } from "@storybook/react";
import Accordion, { AccordionItem } from ".";
import { Text } from "../Text";

export default {
  component: Accordion,
  title: "Components/Accordion",
  argTypes: {
    summary: {
      control: {
        type: "text",
      },
    },
    details: {
      control: {
        type: "text",
      },
    },
  },
};

const Template: ComponentStory<typeof Accordion> = () => (
  <Accordion>
    <AccordionItem summary="Lorem ipsum dolor sit amet">
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam.
      </Text>
    </AccordionItem>
    <AccordionItem summary="Ut enim ad minim veniam.">
      <img alt="foo" src="//source.unsplash.com/user/erondu/200x200" />
    </AccordionItem>
    <AccordionItem summary="Lorem ipsum dolor sit amet">
      <Text>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
        tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim
        veniam.
      </Text>
    </AccordionItem>
  </Accordion>
);

export const Default = Template.bind({});
