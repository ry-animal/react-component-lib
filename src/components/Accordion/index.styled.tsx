import styled from "styled-components";

export const AccordionWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const AccordionItemWrapper = styled.div`
  transition: background-color 0.5s ease-in-out;
  ${({ theme }) => `
    border-bottom: 1px solid ${theme.color.SURFACE_3};
    &:hover, &:focus {
      background-color: ${theme.color.SURFACE_2};
    }
  `}
`;

export const SummaryWrapper = styled.button`
  display: flex;
  align-items: center;
  background: transparent;
  border: none;
  width: 100%;
  padding: ${({ theme }) => `${theme.space.s24} ${theme.space.s16}`};
  cursor: pointer;

  &:focus-visible {
    outline: 1px solid ${({ theme }) => theme.color.SURFACE_4};
  }
`;

export const SummaryText = styled.div`
  margin-left: ${({ theme }) => theme.space.s8};
`;

export const AccordionIcon = styled.span<{ active: boolean }>`
  margin-top: ${({ theme }) => theme.space.s8};
  transform: rotate(${({ active }) => (active ? "90deg" : "0")});
  transform-origin: 60% 40%;
  transition: transform 0.25s ease-in-out;
`;

export const DetailsWrapper = styled.div<{ active: boolean }>`
  display: ${({ active }) => (active ? "block" : "none")};
  transition: auto 0.5s ease-in-out;
  overflow: hidden;
  padding: ${({ theme }) => `${theme.space.s4} ${theme.space.s16}`};
`;
