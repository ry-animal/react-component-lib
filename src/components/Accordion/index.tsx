import * as React from "react";
import { Children, ReactElement, useState } from "react";
import { IconChevronRight } from "../Icons";
import { Text } from "../Text";
import {
  AccordionIcon,
  AccordionItemWrapper,
  AccordionWrapper,
  DetailsWrapper,
  SummaryText,
  SummaryWrapper,
} from "./index.styled";

type AccordionItem = {
  summary: string;
  children: React.ReactNode | React.ReactNode[];
  active?: boolean;
  iconAlt?: string;
  onClick?: () => void;
};

export function AccordionItem({
  summary,
  children,
  active,
  onClick,
}: AccordionItem) {
  return (
    <AccordionItemWrapper>
      <SummaryWrapper onClick={onClick}>
        <AccordionIcon active={active || false}>
          <IconChevronRight size={20} />
        </AccordionIcon>
        <SummaryText>
          <Text text={summary} variant="Body1" align="left" reset />
        </SummaryText>
      </SummaryWrapper>
      <DetailsWrapper active={active || false}>{children}</DetailsWrapper>
    </AccordionItemWrapper>
  );
}

type AccordionProps = {
  children: ReactElement[];
};

export function Accordion({ children }: AccordionProps) {
  const [activeKey, setActiveKey] = useState<number | null>(null);

  return (
    <AccordionWrapper>
      {Children.map(children, (child, index) =>
        React.cloneElement(child, {
          onClick: () => {
            setActiveKey(index === activeKey ? null : index);
          },
          key: index,
          active: index === activeKey,
        })
      )}
    </AccordionWrapper>
  );
}

export default Accordion;
