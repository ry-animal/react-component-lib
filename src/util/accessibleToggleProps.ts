export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
