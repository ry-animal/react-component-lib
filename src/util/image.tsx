import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
