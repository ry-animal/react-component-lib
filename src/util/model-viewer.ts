/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
