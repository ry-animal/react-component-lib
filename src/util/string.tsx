const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
