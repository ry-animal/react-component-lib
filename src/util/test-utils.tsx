import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
