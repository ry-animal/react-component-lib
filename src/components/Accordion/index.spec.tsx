import userEvent from "@testing-library/user-event";
import Accordion, { AccordionItem } from ".";
import { render } from "../../util/test-utils";

describe("<Accordion />", () => {
  describe("rendering", () => {
    test("should match snapshot", () => {
      expect(
        render(
          <Accordion>
            <AccordionItem summary="gm">
              <div>fren</div>
            </AccordionItem>
            <AccordionItem summary="gn">
              <div>moon</div>
            </AccordionItem>
          </Accordion>
        ).asFragment()
      ).toMatchSnapshot();
    });

    it("should only render one at a time", () => {
      const result = render(
        <Accordion>
          <AccordionItem summary="gm">
            <div>fren</div>
          </AccordionItem>
          <AccordionItem summary="gn">
            <div>moon</div>
          </AccordionItem>
        </Accordion>
      );

      const accordianButton = (text: string) => result.getByText(text);

      userEvent.click(accordianButton("gm"));
      expect(result.queryByText("fren")).toBeVisible();

      userEvent.click(accordianButton("gn"));
      expect(result.queryByText("fren")).not.toBeVisible();

      userEvent.click(accordianButton("gm"));
      expect(result.queryByText("moon")).not.toBeVisible();
    });

    it("should close itself if clicked again", () => {
      const result = render(
        <Accordion>
          <AccordionItem summary="gm">
            <div>fren</div>
          </AccordionItem>
          <AccordionItem summary="gn">
            <div>moon</div>
          </AccordionItem>
        </Accordion>
      );

      const accordianButton = (text: string) => result.getByText(text);

      userEvent.click(accordianButton("gm"));
      expect(result.queryByText("fren")).toBeVisible();

      userEvent.click(accordianButton("gm"));
      expect(result.queryByText("moon")).not.toBeVisible();
    });
  });
});
