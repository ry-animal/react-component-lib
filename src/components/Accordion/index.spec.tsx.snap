// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> rendering should match snapshot 1`] = `
<DocumentFragment>
  .c5 {
  color: #151515;
  margin: 0;
  text-align: left;
  font-family: 'Poppins',sans-serif;
  font-weight: 400;
  font-size: 18px;
  line-height: 28px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  -webkit-transition: background-color 0.5s ease-in-out;
  transition: background-color 0.5s ease-in-out;
  border-bottom: 1px solid #EEEEEE;
}

.c1:hover,
.c1:focus {
  background-color: #FFFFFF;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border: none;
  width: 100%;
  padding: 24px 16px;
  cursor: pointer;
}

.c2:focus-visible {
  outline: 1px solid #DBDBDB;
}

.c4 {
  margin-left: 8px;
}

.c3 {
  margin-top: 8px;
  -webkit-transform: rotate(0);
  -ms-transform: rotate(0);
  transform: rotate(0);
  -webkit-transform-origin: 60% 40%;
  -ms-transform-origin: 60% 40%;
  transform-origin: 60% 40%;
  -webkit-transition: -webkit-transform 0.25s ease-in-out;
  -webkit-transition: transform 0.25s ease-in-out;
  transition: transform 0.25s ease-in-out;
}

.c6 {
  display: none;
  -webkit-transition: auto 0.5s ease-in-out;
  transition: auto 0.5s ease-in-out;
  overflow: hidden;
  padding: 4px 16px;
}

@media (max-width:767px) {
  .c5 {
    font-size: 16px;
    line-height: 24px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        <span
          class="c3"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 26 26"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.2197 18.2803C9.92678 17.9874 9.92678 17.5126 10.2197 17.2197L14.6893 12.75L10.2197 8.28033C9.92678 7.98744 9.92678 7.51256 10.2197 7.21967C10.5126 6.92678 10.9874 6.92678 11.2803 7.21967L16.2803 12.2197C16.5732 12.5126 16.5732 12.9874 16.2803 13.2803L11.2803 18.2803C10.9874 18.5732 10.5126 18.5732 10.2197 18.2803Z"
              fill="#151515"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="c4"
        >
          <p
            class="c5"
          >
            gm
          </p>
        </div>
      </button>
      <div
        class="c6"
      >
        <div>
          fren
        </div>
      </div>
    </div>
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        <span
          class="c3"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 26 26"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10.2197 18.2803C9.92678 17.9874 9.92678 17.5126 10.2197 17.2197L14.6893 12.75L10.2197 8.28033C9.92678 7.98744 9.92678 7.51256 10.2197 7.21967C10.5126 6.92678 10.9874 6.92678 11.2803 7.21967L16.2803 12.2197C16.5732 12.5126 16.5732 12.9874 16.2803 13.2803L11.2803 18.2803C10.9874 18.5732 10.5126 18.5732 10.2197 18.2803Z"
              fill="#151515"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <div
          class="c4"
        >
          <p
            class="c5"
          >
            gn
          </p>
        </div>
      </button>
      <div
        class="c6"
      >
        <div>
          moon
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
